### PR TITLE
Fixed table difference insert during price reindex

### LINKED ISSUE #36471: catalog_product_price index getting stuck
### Preconditions and environment

- Magento 2.4.5-p1
- Large catalog (~8,000 products)
- All indexes set to "UPDATE BY SCHEDULE"

### Steps to reproduce

Let indexes run on cron

### Expected result

Indexes run successfully

### Actual result

* Indexes get stuck at `catalog_product_price`.
* Product disappear from categories

Logs show the following errors:

```
[2022-10-13T09:59:05.759670+00:00] main.ERROR: Cron Job indexer_update_all_views has an error: SQLSTATE[21S01]: Insert value list does not match column list:
1136 Column count doesn't match value count at row 1, query was: INSERT INTO `catalog_product_index_price` SELECT `ip_tmp`.* FROM `catalog_product_index_price_temp` AS `ip_tmp` ON DUPLICATE KEY UPDATE `tax_class_id` = VALUES(`tax_class_id`), `price` = VALUES(`price`), `final_price` = VALUES(`final_price`), `min_price` = VALUES(`min_price`), `max_price` = VALUES(`max_price`), `tier_price` = VALUES(`tier_price`). Statistics: {"sum":0,"count":1,"realmem":0,"emalloc":0,"realmem_start":256376832,"emalloc_start":237979584} [] []
[2022-10-13T09:59:05.760119+00:00] main.CRITICAL: PDOException: SQLSTATE[21S01]: Insert value list does not match column list: 1136 Column count doesn't match value count at row 1 in /vendor/magento/framework/DB/Statement/Pdo/Mysql.php:90
Stack trace:
#0 /vendor/magento/framework/DB/Statement/Pdo/Mysql.php(90): PDOStatement->execute()
#1 /vendor/magento/framework/DB/Statement/Pdo/Mysql.php(106): Magento\Framework\DB\Statement\Pdo\Mysql->Magento\Framework\DB\Statement\Pdo\{closure}()
#2 /vendor/magento/framework/DB/Statement/Pdo/Mysql.php(91): Magento\Framework\DB\Statement\Pdo\Mysql->tryExecute(Object(Closure))
#3 /vendor/magento/zendframework1/library/Zend/Db/Statement.php(313): Magento\Framework\DB\Statement\Pdo\Mysql->_execute(Array)
#4 /vendor/magento/zendframework1/library/Zend/Db/Adapter/Abstract.php(480): Zend_Db_Statement->execute(Array)
#5 /vendor/magento/zendframework1/library/Zend/Db/Adapter/Pdo/Abstract.php(247): Zend_Db_Adapter_Abstract->query('INSERT INTO `ca...', Array)
#6 /vendor/magento/framework/DB/Adapter/Pdo/Mysql.php(564): Zend_Db_Adapter_Pdo_Abstract->query('INSERT INTO`ca...', Array)
#7 /vendor/magento/framework/DB/Adapter/Pdo/Mysql.php(634): Magento\Framework\DB\Adapter\Pdo\Mysql->_query('INSERT INTO `ca...', Array)
#8 /vendor/magento/module-catalog/Model/Indexer/Product/Price/AbstractAction.php(193): Magento\Framework\DB\Adapter\Pdo\Mysql->query('INSERT INTO `ca...')
#9 /vendor/magento/module-catalog/Model/Indexer/Product/Price/AbstractAction.php(417): Magento\Catalog\Model\Indexer\Product\Price\AbstractAction->_syncData(Array)
#10 /vendor/magento/module-catalog/Model/Indexer/Product/Price/Action/Rows.php(126): Magento\Catalog\Model\Indexer\Product\Price\AbstractAction->_reindexRows(Array)
#11 /vendor/magento/module-catalog/Model/Indexer/Product/Price.php(68): Magento\Catalog\Model\Indexer\Product\Price\Action\Rows->execute(Array)
#12 /vendor/magento/framework/Interception/Interceptor.php(58): Magento\Catalog\Model\Indexer\Product\Price->execute(Array)
#13 /vendor/magento/framework/Interception/Interceptor.php(138): Magento\Catalog\Model\Indexer\Product\Price\Interceptor->___callParent('execute', Array)

Next Zend_Db_Statement_Exception: SQLSTATE[21S01]: Insert value list does not match column list: 1136 Column count doesn't match value count at row 1, query was: INSERT INTO `catalog_product_index_price` SELECT `ip_tmp`.* FROM `catalog_product_index_price_temp` AS `ip_tmp` ON DUPLICATE KEY UPDATE `tax_class_id` = VALUES(`tax_class_id`), `price` = VALUES(`price`), `final_price` = VALUES(`final_price`), `min_price` = VALUES(`min_price`), `max_price` = VALUES(`max_price`), `tier_price` = VALUES(`tier_price`) in /vendor/magento/framework/DB/Statement/Pdo/Mysql.php:109
Stack trace:
#0 /vendor/magento/framework/DB/Statement/Pdo/Mysql.php(91): Magento\Framework\DB\Statement\Pdo\Mysql->tryExecute(Object(Closure))
#1 /vendor/magento/zendframework1/library/Zend/Db/Statement.php(313): Magento\Framework\DB\Statement\Pdo\Mysql->_execute(Array)
#2 /vendor/magento/zendframework1/library/Zend/Db/Adapter/Abstract.php(480): Zend_Db_Statement->execute(Array)
#3 /vendor/magento/zendframework1/library/Zend/Db/Adapter/Pdo/Abstract.php(247): Zend_Db_Adapter_Abstract->query('INSERT INTO `ca...', Array)
#4 /vendor/magento/framework/DB/Adapter/Pdo/Mysql.php(564): Zend_Db_Adapter_Pdo_Abstract->query('INSERT INTO`ca...', Array)
#5 /vendor/magento/framework/DB/Adapter/Pdo/Mysql.php(634): Magento\Framework\DB\Adapter\Pdo\Mysql->_query('INSERT INTO `ca...', Array)
#6 /vendor/magento/module-catalog/Model/Indexer/Product/Price/AbstractAction.php(193): Magento\Framework\DB\Adapter\Pdo\Mysql->query('INSERT INTO `ca...')
#7 /vendor/magento/module-catalog/Model/Indexer/Product/Price/AbstractAction.php(417): Magento\Catalog\Model\Indexer\Product\Price\AbstractAction->_syncData(Array)
#8 /vendor/magento/module-catalog/Model/Indexer/Product/Price/Action/Rows.php(126): Magento\Catalog\Model\Indexer\Product\Price\AbstractAction->_reindexRows(Array)
#9 /vendor/magento/module-catalog/Model/Indexer/Product/Price.php(68): Magento\Catalog\Model\Indexer\Product\Price\Action\Rows->execute(Array)
#10 /vendor/magento/framework/Interception/Interceptor.php(58): Magento\Catalog\Model\Indexer\Product\Price->execute(Array)
#11 /vendor/magento/framework/Interception/Interceptor.php(138): Magento\Catalog\Model\Indexer\Product\Price\Interceptor->___callParent('execute', Array)
```

Changing `catalog_product_price` index to "Update on Save" seems to work as a temporary workaround.

### Additional information

_No response_

### Release note

_No response_

### Triage and priority

- [ ] Severity: **S0** _- Affects critical data or functionality and leaves users without workaround._
- [X] Severity: **S1** _- Affects critical data or functionality and forces users to employ a workaround._
- [ ] Severity: **S2** _- Affects non-critical data or functionality and forces users to employ a workaround._
- [ ] Severity: **S3** _- Affects non-critical data or functionality and does not force users to employ a workaround._
- [ ] Severity: **S4** _- Affects aesthetics, professional look and feel, “quality” or “usability”._

### PR DESCRIPTION
**Pull request by ilnytskyi**
### Description (*)
Problem appears when dimensions are not used.
`catalog_product_index_price` is now incompatible with `catalog_product_index_price_tmp` as new ID columnt added
https://devdocs.magento.com/guides/v2.4/release-notes/backward-incompatible-changes/reference.html
so method works incorrectly as final query contains additional column
```
\Magento\Catalog\Model\Indexer\Product\Price\AbstractAction::_syncData
```

We faced this issue while running graphql tests when indexers is set - `update on save`
The issue might be skipped in fixtures but when tests are placing the order - it triggers reindex that blocks order placing.

In query logs we may see how temp table is created
```
## 140 ## QUERY
SQL: CREATE TEMPORARY TABLE IF NOT EXISTS `catalog_product_index_price_temp` LIKE `catalog_product_index_price_tmp`
AFF: 0
TIME: 0.0008
```

### Fixed Issues (if relevant)

cannot reindex price 
```
Magento\Framework\Exception\LocalizedException: SQLSTATE[21S01]: Insert value list does not match column list: 1136 Column count doesn't match value count at row 1, query was: INSERT INTO `catalog_product_index_price` SELECT `ip_tmp`.* FROM `catalog_product_index_price_temp` AS `ip_tmp` ON DUPLICATE KEY UPDATE `tax_class_id` = VALUES(`tax_class_id`), `price` = VALUES(`price`), `final_price` = VALUES(`final_price`), `min_price` = VALUES(`min_price`), `max_price` = VALUES(`max_price`), `tier_price` = VALUES(`tier_price`) in /home/www/magento/vendor/magento/module-catalog/Model/Indexer/Product/Price/Action/Row.php on line 32
```
1. Fixes magento/magento2#<issue_number>

### Manual testing scenarios (*)

0. Set indexer update on save
1. Create product on 2.4.5 installation
2. Create custom script to reindex
3. RUN `Magento\Catalog\Model\Indexer\Product\Price::executeRow`
4. See error.
5. Apply patch from this PR
6. retry reindex - no error

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] README.md files for modified modules are updated and included in the pull request if any [README.md predefined sections](https://github.com/magento/devdocs/wiki/Magento-module-README.md) require an update
 - [ ] All automated tests passed successfully (all builds are green)

### Resolved issues:
1. [x] resolves magento/magento2#36471: Fixed table difference insert during price reindex

### CODE REVIEW

**Comment by ilnytskyi** [comment]
Selection_132
@kandy could you take a look?
As I see in logs incompatibility beetween `catalog_product_index_price_tmp` and `catalog_product_index_price` was introduced in your commit.

**Comment by Den4ik** [comment]
@magento run all tests

**Comment by Den4ik** [comment]
@magento run all tests

**Comment by Den4ik** [comment]
@magento run Functional Tests EE, Functional Tests B2B

**Comment by Den4ik** [comment]
@magento run Functional Tests B2B

**Comment by Den4ik** [comment]
https://github.com/magento run Functional Tests B2B

**Comment by Den4ik** [comment]
@magento run Functional Tests B2B

**Review by Den4ik — APPROVED** [review]
✅ Approved
Failed B2B test is not related to PR

File: app/code/Magento/Catalog/Model/Indexer/Product/Price/AbstractAction.php [hunk 1]
```diff
@@ -179,7 +179,8 @@ protected function _syncData(array $processIds = [])
         // for backward compatibility split data from old idx table on dimension tables
         foreach ($this->dimensionCollectionFactory->create() as $dimensions) {
             $insertSelect = $this->getConnection()->select()->from(
-                ['ip_tmp' => $this->_defaultIndexerResource->getIdxTable()]
+                ['ip_tmp' => $this->_defaultIndexerResource->getIdxTable()],
+                array_keys($this->getConnection()->describeTable('catalog_product_index_price'))
```
Comment (kandy) [inline app/code/Magento/Catalog/Model/Indexer/Product/Price/AbstractAction.php]:
```suggestion
                array_keys($this->getConnection()->describeTable(this->tableMaintainer->getMainTableByDimensions($dimensions)))
```

**Review by michaellehmkuhl — COMMENTED** [review]
I believe `this->tableMaintainer->getMainTableByDimensions($dimensions)` was intended to use `$this` rather than `this`.

**Comment by Den4ik** [comment]
@magento run Integration Tests, Functional Tests EE, Functional Tests B2B

File: app/code/Magento/Catalog/Model/Indexer/Product/Price/AbstractAction.php (hunk 1 above)
Comment (skovalenk) [reply app/code/Magento/Catalog/Model/Indexer/Product/Price/AbstractAction.php]: Can you explain please to me, why are so complicating things and could not just align 2 tables together?

**Comment by Den4ik** [comment]
@magento run Functional Tests B2B

**Comment by engcom-Lima** [comment]
@magento run Functional Tests B2B

**Comment by engcom-Echo** [comment]
@ilnytskyi @Den4ik @kandy @michaellehmkuhl 
Thanks for the collaboration & contribution!


:heavy_check_mark: QA Passed
Fixed table difference insert during price reindex 

**Manual Scenario Steps**

1. Put the [module](https://github.com/magento/magento2/files/10235537/Adobe.zip) in `app\code` folder.
2. Run bin/magento setup:upgrade command
3. Create a new product of the new Product type created by the above module.
4. Indexer should be on "Update by Schedule"
5. Run the below command to trigger the error:
6. n98-magerun2 sys:cron:run indexer_update_all_views

**Before:** :heavy_multiplication_x: 
image


**After:** :heavy_check_mark: 
image

**Comment by engcom-Echo** [comment]
Functional Test B2B failures are different on last two run 
Run1
image

Run2
image

Hence moving to `Merge In Progress`.

**Comment by engcom-Echo** [comment]
@magento run Functional Tests B2B

**Comment by maaarghk** [comment]
why did this not make it into 2.4.6 : (

**Comment by EclipseEternal** [comment]
Hi, do we have an idea when will this be available or in which version?

**Comment by TuVanDev** [comment]
@EclipseEternal This pull request has been merged into Magento 2.4.7-beta1 and 2.4.7-beta2. As a result, when Magento 2.4.7 is released next April, it will include version 2.4.7.
To identify the branches where the commits from this pull request were merged, you can refer to the git commit. For example, https://github.com/magento/magento2/pull/36370/commits/d53d0f9c1ba80e5914be687c4a26735d1d807853
For more details on future release information, refer to the release schedule https://experienceleague.adobe.com/docs/commerce-operations/release/planning/schedule.html?lang=en

**Comment by EclipseEternal** [comment]
Hi, thank you for your tips, we will patch then waiting for the 2.4.7.

**Pull request merged**